include/openthread/tcp.h

```diff
@@ -0,0 +1,616 @@
+
+/**
+ * This structure represents a TCP endpoint.
+ *
+ * An TCP endpoint acts an endpoint of TCP connection. It can be used to
+ * initiate TCP connections, and, once a TCP connection is established, send
+ * data to and receive data from the connection peer.
+ */
+typedef struct otTcpEndpoint
+{
+    struct otTcpEndpoint *mNext;                      ///< A pointer to the next TCP endpoint (internal use only)
+    void *mContext;                                   ///< A pointer to application-specific context
+    otTcpEstablished mEstablishedCallback;            ///< "Established" callback function
+    otTcpSendDone mSendDoneCallback;                  ///< "Send done" callback function
+    otTcpSendReady mSendReadyCallback;                ///< "Send ready" callback function
+    otTcpReceiveAvailable mReceiveAvailableCallback;  ///< "Receive available" callback function
+    otTcpDisconnected mDisconnectedCallback;          ///< "Disconnected" callback function
+
+    /* Other implementation-defined fields go here. */
+} otTcpEndpoint;
+
+/**
+ * This structure contains arguments to the @p otTcpEndpointInitialize function.
+ */
+typedef struct otTcpEndpointInitializeArgs
+{
+    otTcpEstablished mEstablishedCallback;            ///< "Established" callback function
+    otTcpSendDone mSendDoneCallback;                  ///< "Send done" callback function
+    otTcpBytesAcked mBytesAckedCallback;              ///< "Bytes acked" callback
+    otTcpSendReady mSendReadyCallback;                ///< "Send ready" callback function
+    otTcpReceiveAvailable mReceiveAvailableCallback;  ///< "Receive available" callback function
+    otTcpDisconnected mDisconnectedCallback;          ///< "Disconnected" callback function
+    void *mContext;                                   ///< Pointer to application-specific context
+    void *mReceiveBuffer;                             ///< Pointer to memory provided to the system for the TCP receive buffer
+    size_t mReceiveBufferSize;                        ///< Size of memory provided to the system for the TCP receive buffer
+} otTcpEndpointInitializeArgs;
+
+/**
+ * Initializes a TCP endpoint. Calling this function causes OpenThread to keep
+ * track of the TCP endpoint and store and retrieve TCP data inside the
+ * @p aEndpoint. The application should refrain from directly accessing or
+ * modifying the fields in @p aEndpoint. If the application needs to reclaim
+ * the memory backing @p aEndpoint, it should call @p otTcpEndpointDeinitialize.
+ *
+ * @param[in]  aInstance  A pointer to an OpenThread instance.
+ * @param[in]  aEndpoint  A pointer to a TCP endpoint structure.
+ * @param[in]  aArgs      A pointer to a structure of arguments.
+ * @param[in]  aContext   A pointer to application-specific context.
+ *
+ * @retval OT_ERROR_NONE    Successfully opened the TCP endpoint.
+ * @retval OT_ERROR_FAILED  Failed to open the TCP endpoint.
```

Comment:
In this case, I actually can't think of a reason why `otTcpEndpointInitialize` would fail. I'm keeping it in the API in case that changes in the future, but for now, the implementation I have in mind would always return `kErrorNone`. So I can't really clarify when the operation would fail.